components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -8588,9 +8646,24 @@ public boolean isSharedScopeNameExists(String scopeName, String tenantDomain) th
     @Override
     public String addSharedScope(Scope scope, String tenantDomain) throws APIManagementException {
 
-        KeyManagerHolder.getKeyManagerInstance(tenantDomain).registerScope(scope);
-        if (log.isDebugEnabled()) {
-            log.debug("Adding shared scope mapping: " + scope.getKey());
+        Set<Scope> scopeSet = new HashSet<>();
+        scopeSet.add(scope);
+        int tenantId = APIUtil.getTenantIdFromTenantDomain(tenantDomain);
+        scopesDAO.addScopes(scopeSet, tenantId);
+        Map<String, KeyManagerDto> tenantKeyManagers = KeyManagerHolder.getTenantKeyManagers(tenantDomain);
+        for (Map.Entry<String, KeyManagerDto> keyManagerDtoEntry : tenantKeyManagers.entrySet()) {
+            KeyManager keyManager = keyManagerDtoEntry.getValue().getKeyManager();
+            if (keyManager != null) {
+                try{
+                    keyManager.registerScope(scope);
+                }catch (APIManagementException e){
+                    log.error("Error occured while registering Scope in KeyManager " + keyManagerDtoEntry.getKey(), e);
```

Comment:
shouldnt we throw the exception?

Reply:
no since we have multiple keymanagers we only log the exception

Reply:
okay. got it (Y)

Reply:
we can store published key managers in database, @bhathiya ?